backend/api/Areas/Keycloak/Controllers/GroupController.cs

```diff
@@ -0,0 +1,165 @@
+using System;
+using System.Threading.Tasks;
+using AutoMapper;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.Logging;
+using Pims.Keycloak;
+using Pims.Dal.Services.Admin;
+using Entity = Pims.Dal.Entities;
+using KModel = Pims.Keycloak.Models;
+using Model = Pims.Api.Areas.Keycloak.Models;
+using Pims.Api.Policies;
+using Pims.Dal.Security;
+using System.Linq;
+using System.Collections.Generic;
+
+namespace Pims.Api.Areas.Keycloak.Controllers
+{
+    /// <summary>
+    /// GroupController class, provides endpoints for managing groups within keycloak.
+    /// </summary>
+    [HasPermission(Permissions.SystemAdmin)]
+    [ApiController]
+    [Area("keycloak")]
+    [Route("/api/[area]/groups")]
+    public class GroupController : ControllerBase
+    {
+        #region Variables
+        private readonly ILogger<GroupController> _logger;
+        private readonly IPimsAdminService _pimsAdminService;
+        private readonly IMapper _mapper;
+        private readonly IKeycloakAdmin _keycloakAdmin;
+        #endregion
+
+        #region Constructors
+        /// <summary>
+        /// Creates a new instance of a GroupController class.
+        /// </summary>
+        /// <param name="logger"></param>
+        /// <param name="pimsAdminService"></param>
+        /// <param name="mapper"></param>
+        /// <param name="keycloakAdmin"></param>
+        public GroupController(ILogger<GroupController> logger, IPimsAdminService pimsAdminService, IMapper mapper, IKeycloakAdmin keycloakAdmin)
+        {
+            _logger = logger;
+            _pimsAdminService = pimsAdminService;
+            _mapper = mapper;
+            _keycloakAdmin = keycloakAdmin;
+        }
+        #endregion
+
+        #region Endpoints
+        /// <summary>
+        /// Sync keycloak groups into PIMS roles.
+        /// </summary>
+        /// <returns></returns>
+        [HttpPost("sync")]
+        public async Task<IActionResult> SyncGroupsAsync()
+        {
+            var gcount = await _keycloakAdmin.GetGroupCountAsync();
+
+            var groups = new List<KModel.GroupModel>();
+            for (var i = 0; i < gcount; i += 10)
+            {
+                var kgroups = await _keycloakAdmin.GetGroupsAsync(i, 10);
+
+                foreach (var kgroup in kgroups)
+                {
+                    var erole = _pimsAdminService.Role.Find(kgroup.Id);
+
+                    if (erole == null)
+                    {
+                        // Need to add the group as a role within PIMS.
+                        erole = _mapper.Map<Entity.Role>(kgroup);
+                        _pimsAdminService.Role.AddOne(erole);
+                    }
+                    else
+                    {
+                        _mapper.Map(kgroup, erole);
+                        _pimsAdminService.Role.UpdateOne(erole);
+                    }
+
+                    groups.Add(kgroup);
+                }
+
+                _pimsAdminService.CommitTransaction();
```

Comment:
it feels weird to be managing transactions in a controller, is the long term plan to move these to a KeycloakService?

Reply:
Agree, controllers should be thin and delegate to business logic to services.